Modules/_sqlite/clinic/cursor.c.h

```diff
@@ -138,7 +138,7 @@ pysqlite_cursor_fetchone(pysqlite_Cursor *self, PyObject *Py_UNUSED(ignored))
 }
 
 PyDoc_STRVAR(pysqlite_cursor_fetchmany__doc__,
-"fetchmany($self, /, size=cursor.arraysize)\n"
+"fetchmany($self, /, size=<unrepresentable>)\n"
```

Comment:
Unfortunately, `<unrepresentable>` is not the ideal value here. We usually use `None` or `NULL` if we somehow cannot represent the default value in the codebase. We can then document that the default size is `cursor.arraysize` in the docstring. I'd prefer to stay consistent with that approach.

Reply:
Also, default value of arraysize is documented as 1 at https://docs.python.org/3/library/sqlite3.html#sqlite3.Cursor.arraysize Any reason not to use 1 and then document that size is set to cursor.arraysize in the docstring?

Reply:
I don't see how that makes for a simpler solution. This solution ensures (with minimal changes) that the docstring looks ok and it preserves the behaviour. I'll give it a shot, though.

Reply:
Setting `py_default` to `'1'` gives me the following signature:

```
fetchmany(self, /, size=1)
```

Reply:
We should _not_ set `py_default` to `1`. The point of using `<unrepresentable>` is to leave the docstring alone. I've pushed a commit that changes the converter from int to object. A side effect of this is that `None` is now accepted as an argument to `fetchmany`. That is a change in behaviour. We also have to handle the default argument in `pysqlite_cursor_fetchmany_impl` now, instead of letting AC taking care of it.

IMHO, the initial PR is a more elegant solution.

Reply:
The last part of the AC line (`= 1`) only serves to tell AC that this argument has a default argument. `1` is actually never used; if it was, unit tests would fail at `CursorTests.test_array_size`.

Reply:
I've added 3d1616a as an attempt to use the object converter instead.

Reply:
That's interesting, because tests passed in my computer:

```
$ ./python.exe -m test test_sqlite
0:00:00 load avg: 2.27 Run tests sequentially
0:00:00 load avg: 2.27 [1/1] test_sqlite

== Tests result: SUCCESS ==

1 test OK.

Total duration: 451 ms
Tests result: SUCCESS
```

And from the REPL:

```
$ ./python.exe
...
>>> cu.execute("select * from projects").fetchmany()
[(1, 'foo')]
>>> cu.execute("select * from projects").fetchmany(2)
[(1, 'foo'), (2, 'bar')]
>>> cu.arraysize = 3
>>> cu.execute("select * from projects").fetchmany()
[(1, 'foo'), (2, 'bar'), (3, 'baz')]
>>> cu.execute("select * from projects")
<sqlite3.Cursor object at 0x10a1fea40>
>>> cu.fetchmany()
[(1, 'foo'), (2, 'bar'), (3, 'baz')]
```

Reply:
Ah, wait, you said *would*. Meh, I need sleep :) I'll look at this tomorrow morning.

Reply:
Hm, I guess `c_default` overrides then. In any case it still modifies the docstring, which `<unrepresentable>` doesn't.

Reply:
What's wrong with the modified docs, by the way? It would say 1 instead of cursor.arraysize, which is basically equivalent to 1. We can clarify that in the docstring by mentioning that the default value is actually comes from cursor.arraysize, which is perfectly fine to me.

Reply:
> It would say 1 instead of cursor.arraysize, which is basically equivalent to 1.

Not if the user touched it :)

```
import sqlite3
cx = sqlite3.connect(":memory:")
cu = cx.cursor()
cu.arraysize = 2
# The default is now 2, not 1
cu.fetchmany()
```

How about this? I revert 3d1616a and I add a docstring that explains the default value?

Reply:
Python 3.10.0a4 (pre #24007):

```
fetchmany(...)
    Fetches several rows from the resultset.
```

With `int(c_default='self->arraysize') = 1`

```
fetchmany(self, /, size=1)
    Fetches several rows from the resultset.
```

With my proposal:

```
fetchmany(...)
    Fetches several rows from the resultset.
    
    size
      The number of rows to fetch. Defaults to the cursor's arraysize.
```

Reply:
Well, then that would mean they changed the initial *default* value :) That's perfectly fine for me. We have more annoying little discrepancies than this one.

Yeah, let's revert 3d1616a and clarify the docstring.

Reply:
> We have more annoying little discrepancies than this one.

True that! :)

Reply:
I'd suggest:

```
fetchmany(self, /, size=1)
    Fetches several rows from the resultset.

    The default value of size is set by Cursor.arraysize attribute. (wording can be different)
```

Defaukt value is already documented as 1 at https://docs.python.org/3/library/sqlite3.html#sqlite3.Cursor.arraysize

> The default value is 1 which means a single row would be fetched per call.

Reply:
I agree. That's an improvement. Thanks!

Reply:
PTAL!

Reply:
two things:
1. `<unrepresentable>` is still there.
2. You probably forgot to run `make clinic` because the generated file is out of sync :)

Reply:
> 1. `<unrepresentable>` is still there.

All right, I'll remove it, but I don't understand why we can't use it.

> 2. You probably forgot to run `make clinic` because the generated file is out of sync :)

Fixed :)

Reply:
PTAL

Reply:
To clarify, when we set a default value with a constant (e.g. sys.maxsize or os.sep) instead of a direct value (e.g. 9223372036854775807, '/') it usually means that the value can be different depending on the platform. This one is a bit of an edge case and it's OK to have a minor inconsistency *if* a user is changed the default value of cursor.arraysize (I doubt a lot of users know it's possible, to be honest.)

Reply:
Ok, thanks!

Reply:
> All right, I'll remove it, but I don't understand why we can't use it.

It's because I've never seen it's used in Python and it's not entirely correct :) We cannot dynamically represent it now because of a tooling limitation. We can easily represent its initial default value, though.

Reply:
> It's because I've never seen it's used in Python and it's not entirely correct :)

https://github.com/python/cpython/blob/61d8c54f43a7871d016f98b38f86858817d927d5/Modules/posixmodule.c#L13037

Reply:
Thanks, but that's basically what I was trying to explain here :) File descriptor of stdout can be different on different platforms, and since we cannot call sys.stdout.fileno() due to a tooling limitation, and there is no other suitable default value to use.

In our case a) we already have a valid default value b) the attribute is writable by user so its value is dynamic (it's not the case for fd of get_terminal_size())

Reply:
But `<unrepresentable>` only affects the docstring, no?

Reply:
Yes, but then you'll get bug reports about docstring and documentation not in sync. There is no win here, unfortunately.

Reply:
You're probably right, but I'm not convinced :)